### DISCUSSION TITLE
vite-plugin-checker runs multiple times

### DISCUSSION THREAD

**Discussion by mikew** [vikejs/vike#1338] [Discussion]
Repro: https://stackblitz.com/edit/github-twckcs?file=vite.config.ts. This is vike's `react-full-v1` example with:

- vite-plugin-checker added
- eslint added
- only the `no-console` rule enabled

Notice that the eslint results are output multiple times. I'm not sure whether the issue lies in vite-plugin-checker or vike (note I'm still very new to vite's SSR & vike). I can see [vite-plugin-checker added some workaround for this](https://github.com/fi3ework/vite-plugin-checker/pull/203), but it might be better served to check the various `ssr` options vite sets when building SSR.

Workarounds:

Patch vite-plugin-checker so that those `initializaed` and `initializedCounter` are in the topmost scope, that seemed to work on my end.

Or

- set `disableAutoFullBuild: true` in vike
- manually call `vite build --ssr`, and in your `vite.config` check if `process.argv.includes('--ssr')` and don't add `vite-plugin-checker`

**Comment by brillout** [comment]
PR welcome to make sure it runs only once, AFAICT the trick is to remove it from `config.plugins` before [build chaining kicks in](https://vike.dev/disableAutoFullBuild), see `triggerFullBuild()` in `vike/node/plugin/plugins/autoFullBuild.ts`.

**Reply by mikew** [reply]
Happy to PR but there's a few things.

[`configInline` doesn't have the list of plugins](https://github.com/vikejs/vike/blob/b37f10280e86655ec864e78b239bd9c70818226a/vike/node/plugin/plugins/autoFullBuild.ts#L71-L78), but it does have the path to the projects `vite.config`. Passing it through vite's `resolveConfig` like so ...

```diff
-  const configInline = {
+  const configInline = await resolveConfig({
    ...configFromCli,
    configFile: configFromCli?.configFile || config.configFile,
    root: config.root,
    build: {
      ...configFromCli?.build
    }
-  } satisfies InlineConfig
+  } satisfies InlineConfig, 'vike full-build' as any, 'production')
```

... does get the full list of plugins where filtering could happen, but that results in another error:

```
$ ./node_modules/.bin/vite build
...
vite v5.0.6 building SSR bundle for production...
✓ 0 modules transformed.
[commonjs--resolver] The "path" argument must be of type string. Received function assetsInclude
TypeError: The "path" argument must be of type string. Received function assetsInclude
    at new NodeError (node:internal/errors:405:5)
    at validateString (node:internal/validators:162:11)
    at isAbsolute (node:path:1156:5)
    at getMatcherString (file:///home/mike/Work/vike-example-mui-emotion-cache/node_modules/.pnpm/vite@5.0.6_@types+node@20.10.1/node_modules/vite/dist/node/chunks/dep-YJaePtkC.js:2657:37)
    at Object.test (file:///home/mike/Work/vike-example-mui-emotion-cache/node_modules/.pnpm/vite@5.0.6_@types+node@20.10.1/node_modules/vite/dist/node/chunks/dep-YJaePtkC.js:2677:33)
    at result (file:///home/mike/Work/vike-example-mui-emotion-cache/node_modules/.pnpm/vite@5.0.6_@types+node@20.10.1/node_modules/vite/dist/node/chunks/dep-YJaePtkC.js:2698:25)
    at Object.assetsInclude (file:///home/mike/Work/vike-example-mui-emotion-cache/node_modules/.pnpm/vite@5.0.6_@types+node@20.10.1/node_modules/vite/dist/node/chunks/dep-YJaePtkC.js:67298:52)
    at Object.resolveId (file:///home/mike/Work/vike-example-mui-emotion-cache/node_modules/.pnpm/vite@5.0.6_@types+node@20.10.1/node_modules/vite/dist/node/chunks/dep-YJaePtkC.js:16535:25)
    at Object.handler (file:///home/mike/Work/vike-example-mui-emotion-cache/node_modules/.pnpm/vite@5.0.6_@types+node@20.10.1/node_modules/vite/dist/node/chunks/dep-YJaePtkC.js:66588:19)
    at file:///home/mike/Work/vike-example-mui-emotion-cache/node_modules/.pnpm/rollup@4.6.1/node_modules/rollup/dist/es/shared/node-entry.js:18606:40 {
  code: 'PLUGIN_ERROR',
  pluginCode: 'ERR_INVALID_ARG_TYPE',
  plugin: 'commonjs--resolver',
  hook: 'resolveId'
}
```

the next best thing I can think of is setting an env var

```diff
+  process.env.VIKE_SSR = 'true'
  await build({
    ...configInline,
    build: {
      ...configInline.build,
      ssr: true
    }
  })
```

which can then be picked up in the users `vite.config` file and they can do whatever.

**Reply by brillout** [reply]
I think the cleanest would be a new internal global variable (see `getGlobalObject()`) then remove `vite-plugin-checker` accordingly.

Or, atlernatively, pushing a plugin to `configInline` on-the-fly that removes `vite-plugin-checker`.

**Reply by mikew** [reply]
So with the former, there'd be a new global variable accessible in `vite.config` where projects could do whatever they want when it's set? I like that.

I'll still reach out to vite-plugin-checker, as this issue will pop up with any project where vite's `build()` is called multiple times.

Edit: I should mention, I still think there's merit to setting an environment variable, as vite does set `import.meta.env.SSR = 'true'` for SSR builds.

**Comment by brillout** [comment]
Correct if I'm wrong, but I think it's possible to do it automatically
without the user having to do anything.

**Reply by mikew** [reply]
Is that via what you suggested with "pushing a plugin to configInline on-the-fly that removes vite-plugin-checker". I agree automatic would be nice, just worried about:

- it turning into a cat and mouse game with other plugins
- letting the user provide a list of plugins that should be disabled in SSR builds
- vite-plugin-checker renaming and nullifying the automatic removal

**Reply by mikew** [reply]
Just coming back to this idea, and thinking more about

> pushing a plugin to configInline on-the-fly that removes vite-plugin-checker

Is it even possible for plugins to remove other plugins? Looking at the source code for `Plugin.config`, and it seems you can return anything except a modified list of plugins: https://github.com/vitejs/vite/blob/b1a6c84c3b55f9a79f63e1fd9bd11ac9ef151e92/packages/vite/src/node/plugin.ts#L75-L78

**Reply by brillout** [reply]
You can mutate the `plugins` array. Ugly but works.

**Reply by mikew** [reply]
I've tried a couple of ways, the `console.log`s do validate it's not in the array, but it still actually executes. Dunno if this behaviour changed recently with Vite 5.

```ts
{
  name: 'test-remove-plugins',
  enforce: 'pre',
  config(userConfig) {
    const index = userConfig.plugins?.findIndex((x) => {
      if (x === false || x == null) {
        return false
      }

      // TODO Should probably handle these.
      if (isPromise(x)) {
        return false
      }

      // TODO Should probably handle these.
      if (Array.isArray(x)) {
        return false
      }

      return x.name === 'vite-plugin-checker'
    })

    if (index != null && index !== -1) {
      userConfig.plugins?.splice(index, 1)
    }

    console.log(userConfig.plugins)
  },
}
```

```ts
{
  name: 'test-remove-plugins',
  enforce: 'pre',
  config(userConfig) {
    const filtered = userConfig.plugins?.filter((x) => {
      if (x === false || x == null) {
        return false
      }

      // TODO Should probably handle these.
      if (isPromise(x)) {
        return true
      }

      // TODO Should probably handle these.
      if (Array.isArray(x)) {
        return true
      }

      return x.name !== 'vite-plugin-checker'
    })

    console.log(filtered)

    userConfig.plugins = filtered
  },
}
```

I also didn't realize that the config passed to `triggerFullBuild` is the resolved config, so maybe `plugins` can be manipulated there.

**Reply by brillout** [reply]
Yea, I'd try in `configResolved()`. It must be somewhere, as attested by Vite's internal code.

**Comment by brillout** [comment]
I agree a cross-project-PR would be best but, in the meantime, AFAICT it's
fine.
